river/metrics/mutual_info.py

```diff
@@ -0,0 +1,498 @@
+        super().__init__(cm)
+        self.average_method = average_method
+
+    @property
+    def works_with_weights(self):
+        return False
+
+    @staticmethod
+    def _generalized_average(u, v, average_method):
+
+        if average_method == "min":
+            return min(u, v)
+        elif average_method == "max":
+            return max(u, v)
+        elif average_method == "geometric":
+            return math.sqrt(u * v)
+        elif average_method == "arithmetic":
+            return (u + v) / 2
+        else:
+            raise ValueError(
+                "'average_method' must be either 'min', 'max', "
+                "'geometric', or 'arithmetic' "
+            )
+
+    def get(self):
+
+        n_classes = len([i for i in self.cm.sum_row.values() if i != 0])
+        n_clusters = len([i for i in self.cm.sum_col.values() if i != 0])
+
+        if (n_classes == n_clusters == 1) or (n_classes == n_clusters == 0):
+            return 1.0
+
+        mutual_info_score = metrics.MutualInfo(self.cm).get()
+
+        entropy_true = entropy_pred = 0.0
+
+        for i in self.cm.classes:
+
+            try:
+                entropy_true -= (
+                    self.cm.sum_row[i]
+                    / self.cm.n_samples
+                    * math.log(self.cm.sum_row[i] / self.cm.n_samples)
+                )
+            except ValueError:
+                pass
+
+            try:
+                entropy_pred -= (
+                    self.cm.sum_col[i]
+                    / self.cm.n_samples
+                    * math.log(self.cm.sum_col[i] / self.cm.n_samples)
+                )
+            except ValueError:
+                pass
+
+        normalizer = self._generalized_average(
+            entropy_true, entropy_pred, self.average_method
+        )
+
+        normalizer = max(normalizer, np.finfo("float64").eps)
+
+        return mutual_info_score / normalizer
+
+
+class ExpectedMutualInfo(metrics.MultiClassMetric):
+    r"""Expected Mutual Information.
+
+    Like the Rand index, the baseline value of mutual information between two
+    random clusterings is not necessarily a constant value, and tends to become
+    larger when the two partitions have a higher number of cluster (with a fixed
+    number of samples $N$). Using a hypergeometric model of randomness, it can be
+    shown that the expected mutual information between two random clusterings [^1] is:
+
+    $$
+    E(MI(U, V)) = \sum_{i=1}^R \sum_{i=1}^C \sum{n_{ij}=(a_i+b_j-N)^+}^{\min(a_i,b_j)} \frac{n_{ij}}{N} \log(\frac{N n_{ij}}{a_i b_j}) \frac{a_i! b_j! (N-a_i)! (N-b_j)!}{N! n_{ij}! (a_i - n_{ij})! (b_j - n_{ij})! (N - a_i - b_j + n_{ij})!}}
+    $$
+
+    where
+
+    * $(a_i + b_j - N)^+$ denotes $\max(1, a_i + b_j - N)$,
+
+    * $a_i$ is the sum of row i-th of the contingency table, and
+
+    * $b_j$ is the sum of column j-th of the contingency table.
+
+    The Expected Mutual Information score will be used to calculate the Adjusted Mutual
```

Comment:
~~This note should be in the AMI class.~~

Ignore this, I just saw the note in AMI.

Reply:
~~Also, consider the following rewording:

> The Adjusted Mutual Information score (AMI) relies on the Expected Mutual Information 
   (EMI) score. Due to the complexity of EMI, the AMI is one order of magnitude slower than 
    other metrics.~~

Reply:
Replace with a note here mentioning that the metric is very expensive to calculate.

Reply:
Resolved!